osquery/tables/yara/yara.cpp

```diff
@@ -39,71 +40,221 @@ FLAG(bool,
+            disable_yara_string_private,
+            false,
+            "The yara strings are private by default. The flag will disable "
+            "the feature. ");
+
 namespace tables {
 
+typedef enum { YC_NONE = 0, YC_GROUP, YC_FILE, YC_RULE } YaraScanType;
+
+using YaraScanContext = std::set<std::pair<YaraScanType, std::string>>;
+
+static inline bool isNull(std::shared_ptr<ConfigParserPlugin> parser) {
+  return (parser == nullptr) || (parser.get() == nullptr);
+}
+
+static inline std::string hashStr(const std::string& str, YaraScanType yc) {
+  switch (yc) {
+  case YC_RULE:
+    return "rule_" +
+           hashFromBuffer(HASH_TYPE_SHA256, str.c_str(), str.length());
+  default:
+    return str;
+  }
+};
+
+// Get the yara configuration parser
+static std::shared_ptr<YARAConfigParserPlugin> getYaraParser(void) {
+  auto parser = Config::getParser("yara");
+  if (isNull(parser)) {
+    LOG(ERROR) << "YARA config parser plugin has no pointer";
+    return nullptr;
+  }
+
+  std::shared_ptr<YARAConfigParserPlugin> yaraParser = nullptr;
+  try {
+    yaraParser = std::dynamic_pointer_cast<YARAConfigParserPlugin>(parser);
+  } catch (const std::bad_cast& e) {
+    LOG(ERROR) << "Error casting yara config parser plugin";
+    return nullptr;
+  }
+
+  return yaraParser;
+}
+
 void doYARAScan(YR_RULES* rules,
                 const std::string& path,
                 QueryData& results,
-                const std::string& group,
+                YaraScanType yc,
                 const std::string& sigfile) {
-  Row r;
+  Row row;
 
   // These are default values, to be updated in YARACallback.
-  r["count"] = INTEGER(0);
-  r["matches"] = std::string("");
-  r["strings"] = std::string("");
-  r["tags"] = std::string("");
+  row["count"] = INTEGER(0);
+  row["matches"] = SQL_TEXT("");
+  row["strings"] = SQL_TEXT("");
+  row["tags"] = SQL_TEXT("");
+  row["sig_group"] = SQL_TEXT("");
+  row["sigfile"] = SQL_TEXT("");
+  row["sigrule"] = SQL_TEXT("");
 
   // This could use target_path instead to be consistent with yara_events.
-  r["path"] = path;
-  r["sig_group"] = std::string(group);
-  r["sigfile"] = std::string(sigfile);
+  row["path"] = path;
+
+  switch (yc) {
+  case YC_GROUP:
+    row["sig_group"] = SQL_TEXT(sigfile);
+    break;
+  case YC_FILE:
+    row["sigfile"] = SQL_TEXT(sigfile);
+    break;
+  case YC_RULE:
+    row["sigrule"] = SQL_TEXT(sigfile);
+    break;
+  case YC_NONE:
+    break;
+  }
 
   // Perform the scan, using the static YARA subscriber callback.
   int result = yr_rules_scan_file(
-      rules, path.c_str(), SCAN_FLAGS_FAST_MODE, YARACallback, (void*)&r, 0);
+      rules, path.c_str(), SCAN_FLAGS_FAST_MODE, YARACallback, (void*)&row, 0);
   if (result == ERROR_SUCCESS) {
-    results.push_back(std::move(r));
+    results.push_back(std::move(row));
+  }
+}
+
+Status genYaraRuleFromFile(QueryContext& queryContext,
+                           YaraScanContext& scanContext) {
+  auto yaraParser = getYaraParser();
```

Comment:
The method `getYaraParser` may print an error then this function will "bubble up" a similar error but with a loss of precision. We should check that we do not unintentionally emit the same (or similar) error twice. For example, this status is passed up to `genYara`, do we end up printing a status message there too?

I suggest we avoid writing status messages in `getYaraParser` and allow the top-most calling method to choose to print.

Reply:
I removed the status log printing from the top-most calling function and let `getYaraParser` print the error logs. These logs are more useful to debug the specific issues. The top-most function returns if the parser is null.